src/Compilers/CSharp/Test/Semantic/Semantics/StackAllocInitializerTests.cs

```diff
@@ -5,13 +5,74 @@
 using Microsoft.CodeAnalysis.CSharp.Syntax;
 using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
 using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
 using Xunit;
 
 namespace Microsoft.CodeAnalysis.CSharp.UnitTests
 {
     [CompilerTrait(CompilerFeature.StackAllocInitializer)]
     public class StackAllocInitializerTests : CompilingTestBase
     {
+        [Fact, WorkItem(33945, "https://github.com/dotnet/roslyn/issues/33945")]
+        public void RestrictedTypesAllowedInStackalloc()
+        {
+            var comp = CreateCompilationWithMscorlibAndSpan(@"
+ref struct RefS { }
+
+class C
+{
+    unsafe void M()
+    {
+        var x1 = stackalloc RefS[10];
+        var x2 = stackalloc System.TypedReference[10];
+        var x3 = stackalloc System.ArgIterator[10];
+        var x4 = stackalloc System.RuntimeArgumentHandle[10];
+
+        var y1 = new RefS[10];
+        var y2 = new System.TypedReference[10];
+        var y3 = new System.ArgIterator[10];
+        var y4 = new System.RuntimeArgumentHandle[10];
+
+        RefS[] z1 = null;
+        System.TypedReference[] z2 = null;
+        System.ArgIterator[] z3 = null;
+        System.RuntimeArgumentHandle[] z4 = null;
+        _ = z1;
+        _ = z2;
+        _ = z3;
+        _ = z4;
+    }
```

Comment:
Suggestion for other tests here:

```
ref struct RefG<T> { public T field; } 
```

Then add the following cases into the matrix 

```
RefG<int>
RefG<string>
```